### PR TITLE
Modify rule S5273: Expand and adjust for LaYC

### PR DESCRIPTION
**Pull request by pdschbrt**
**Caution:** our analyzer currently does not correctly detect the issue in the first code example in section `Why is this an issue?`.

## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S5273/cfamily/rule.adoc [hunk 1]
```diff
@@ -1,35 +1,117 @@
...
 
-Moreover, ``++strncat++`` always adds a terminating null character at the end of the appended characters so the size argument should be smaller than the size of the destination to let enough space for it.
+[source,cpp]
+----
+#include <stdio.h>
+#include <string.h>
 
+void foo(const char *src) {
+  char dst[10] = {0};
+  strlcpy(dst, src, strlen(src)); // Noncompliant: size of destination should be used.
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S5273/cfamily/rule.adoc:18]: Again we should merge this change with example that we do not detect. We either change it to `src` or leave this PR open and mention ticket.

**Review by tomasz-kaminski-sonarsource — CHANGES_REQUESTED** [review]
Not sure if we should address these comments now. I do not think we should have non-compliant that we do not detect. So we can either adjust them to be less convincing, or leave this open.

File: rules/S5273/cfamily/rule.adoc [hunk 2]
```diff
@@ -1,35 +1,117 @@
...
+----
 
-=== Noncompliant code example
 
-[source,cpp]
+== What is the potential impact?
+
+By using the source buffer's size to determine the size argument for ``++strncat++``, ``++strlcat++`` or ``++strlcpy++``, the program becomes vulnerable to buffer overflows which pose a _serious_ security risk.
+
+Since C does not provide built-in safeguards against buffer overflows, it is prone to this type of vulnerability.
```
[the remark this answers is not in the document]
Comment (pdschbrt) [reply rules/S5273/cfamily/rule.adoc:28]: Let me try to rephrase this a bit.

File: rules/S5273/cfamily/rule.adoc (hunk 1 above)
Comment (pdschbrt) [reply rules/S5273/cfamily/rule.adoc:18]: I'll update the example code to what was originally intended by this rule and create a Jira ticket for improving on this rule.

**Comment by pdschbrt** [comment]
A corresponding Jira ticket has been added and can be found at https://sonarsource.atlassian.net/browse/CPP-4656

File: rules/S5273/cfamily/rule.adoc [hunk 3]
```diff
@@ -1,35 +1,117 @@
...
+Buffer overflows occur when a program writes data beyond the boundaries of a buffer and can lead to memory corruption and potential security vulnerabilities.
+Attackers can exploit buffer overflows to overwrite critical data, execute arbitrary code, or gain unauthorized access to a system.
+To mitigate this risk, developers must carefully manage buffer sizes (, use secure coding practices, and employ techniques like input validation and bounds checking).
```
[the remark this answers is not in the document]
Comment (pdschbrt) [reply rules/S5273/cfamily/rule.adoc:30]: Otherwise, I could add a `Going the extra mile` subsection`. Let me know what you think.

**Comment by tomasz-kaminski-sonarsource** [comment]
I would mentiuon `std::string` and `std::format` in going extra mile for C++.

**Comment by pdschbrt** [comment]
I've added an "extra mile" section and also move some content a bit. What do you think?

**Review by tomasz-kaminski-sonarsource — APPROVED** [review]
I like these additional sections, they look good.

**Pull request merged**